src/tools/ScenarioMeasurement/Startup/WindowsTraceSession.cs

```diff
@@ -0,0 +1,116 @@
+﻿using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Parsers;
+using Microsoft.Diagnostics.Tracing.Session;
+using System.Collections.Generic;
+using System.IO;
+
+
+namespace ScenarioMeasurement
+{
+    class WindowsTraceSession : ITraceSession
+    {
+        private Logger logger;
+        private string traceFilePath;
+        public TraceEventSession KernelSession { get; set; }
+        public TraceEventSession UserSession { get; set; }
+        private Dictionary<TraceSessionManager.KernelKeyword, KernelTraceEventParser.Keywords> kernelKeywords;
+        private Dictionary<TraceSessionManager.ClrKeyword, ClrPrivateTraceEventParser.Keywords> clrKeywords;
+
+        public WindowsTraceSession(string sessionName, string traceName, string traceDirectory, Logger logger)
+        {
+            this.logger = logger;
+
+            string kernelFileName = Path.ChangeExtension(traceName, "perflabkernel.etl");
+            string userFileName = Path.ChangeExtension(traceName, "perflabuser.etl");
+            traceFilePath = Path.Combine(traceDirectory, Path.ChangeExtension(traceName, ".etl"));
+
+            KernelSession = new TraceEventSession(sessionName + "_kernel", Path.Combine(traceDirectory, kernelFileName));
+            UserSession = new TraceEventSession(sessionName + "_user", Path.Combine(traceDirectory, userFileName));
+            InitWindowsKeywordMaps();
+        }
+
+        public void EnableProviders(IParser parser)
+        {
+            // Enable both providers and start the session
+            parser.EnableKernelProvider(this);
+            parser.EnableUserProviders(this);
+        }
+
+        public void Dispose()
+        {
+            KernelSession.Dispose();
+            UserSession.Dispose();
+
+            MergeFiles(KernelSession.FileName, UserSession.FileName, traceFilePath);
+            logger.Log($"Trace Saved to {traceFilePath}");
+        }
+
+        private void MergeFiles(string kernelTraceFile, string userTraceFile, string traceFile)
+        {
+            var files = new List<string>();
+            if (File.Exists(kernelTraceFile))
+            {
+                files.Add(kernelTraceFile);
+            }
+            if (File.Exists(userTraceFile))
+            {
+                files.Add(userTraceFile);
+            }
+            if (files.Count != 0)
```

Comment:
This construct suggests that one of the kernel or user files may be missing. That makes sense to me as there are cases (like the `ProcessTimeParser`) where there will be no user session. There will ALWAYS be a kernel session, though. Nothing really works without it. I think it's reasonable to fail if there's not a kernel session.

Reply:
Okay, so I will change it to throw exception upon no kernel trace file, and remove the merge condition.